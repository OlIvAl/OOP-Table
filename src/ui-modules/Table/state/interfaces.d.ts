/**
 * Декоратор с функционалом инициализации стора букингов
 */
import { IPaginationObject } from '../../../../../lib/Pagination/interfaces';
import { IFinanceActionData, IFinanceActionTotal } from '../../../../../DataLayer/responses';
import { IFinanceActionDetailStateData } from '../../../../../DataLayer/modules/actions/interfaces';
import { actionType, statusType } from '../../../../../@types/actions';

/**
 * Декоратор с функционалом получения списка по букингов
 */
export interface IFinanceActionsDataService {
  /**
   * Флаг загрузки
   */
  loading: boolean;

  actions: IFinanceActionData[];

  total: IFinanceActionTotal | null;
  totalAmount: number;

  init: (params: IBalanceParams) => AsyncGenerator<void>;
}

/**
 * Декоратор с функционалом получения детализации по букингу
 */
export interface IFinanceActionDetailDataService {
  /**
   * Флаг загрузки
   */
  loading: boolean;
  action: IFinanceActionDetailStateData['action'] | null;
  transactions: IFinanceActionDetailStateData['transactions'];
  transactionsTotal: IFinanceActionDetailStateData['transactionsTotal'] | null;
  additionalFields: IFinanceActionDetailStateData['additionalFields'];
  init: (action_id: number, marker: number, currency: string) => AsyncGenerator<void>;
}

/**
 * Объект с полями для вывода списка мета информации букинга
 */
export interface IDataListItem {
  key: string;
  description: string;
}

/**
 * Объект с детальной информацией по букингу
 */
export interface IBookingDetail {
  price: string;
  total_profit: string;
  action_state: statusType;
  marker: string;
  action_state: string;
  created_at: string;
  campaign_id: number;
  campaign_name: string;
  action_id: number;
  metaList: IDataListItem[];
  action_type: actionType | '';
  host: string;
}

/**
 * Объект строки таблицы букингов
 */
export interface IBalanceRow {
  source: string;
  created_at: string;
  updated_at: string;
  description: string;
  action_state: statusType;
  price: string;
  profit_diff: string;
  total_profit: string;
}

/**
 * Объукт строки "итого" в таблице букингов
 */
export interface IBalanceTotalRow {
  source: '';
  created_at: '';
  updated_at: '';
  description: '';
  action_state: '';
  price: string;
  profit_diff: string;
  total_profit: string;
}

/**
 * Объект строки в таблице транзакций
 */
export interface ITransactionRow {
  created_at: string;
  status: statusType;
  price: string;
  profit_diff: string;
  total_profit: string;
}
/**
 * Объукт строки "итого" в таблице транзакций
 */
export interface ITransactionTotalRow {
  created_at: '';
  status: '';
  price: '';
  profit_diff: '';
  total_profit: string;
}

/**
 * Объект данных для фильтрации списка букингов
 */
export interface IBalanceFilterObject {
  period: { from: string; until: string };
  marker: string;
  status: string;
  campaign_id: number;
}

/**
 * Объект с данными для запроса букингов
 * Состоит из данных пагинации и фильтра
 */
export interface IBalanceParams extends Partial<IBalanceFilterObject>, IPaginationObject {}
