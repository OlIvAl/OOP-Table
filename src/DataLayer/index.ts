import { ActionsRequests } from './modules/actions/requests';

export default class DataLayer extends ActionsRequests {}
