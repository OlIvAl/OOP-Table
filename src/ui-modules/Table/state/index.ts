import { action, computed, flow, IReactionDisposer, reaction } from 'mobx';
import { BalanceRowDecorator } from './decorators/BalanceRowDecorator';
import { BalanceTotalRowDecorator } from './decorators/BalanceTotalRowDecorator';
import { ITable } from '../../../lib/TableBuilder/interfaces';
import { IFinanceActionModel } from '../../../store/Actions/interfaces';
import { TableBuilder } from '../../../lib/TableBuilder/TableBuilder';
import { IBalanceParams, IFinanceActionData } from '../../../DataLayer/modules/actions/interfaces';
import { IStoreWithCurrentModel } from '../../../store/Abstract/interfaces';
import { IBalanceRow, IBalanceTotalRow, IFinanceActionsDataService } from './interfaces';
import { FinanceActionsStoreInitDecorator } from './decorators/FinanceActionsStoreInitDecorator';

/**
 * Объект состояния компонента баланса
 */
export class BalanceTableState {
  @computed get loading(): boolean {
    return this.getFinanceActionsService.loading;
  }
  @computed get balanceTableShown(): boolean {
    return !!this.financeActionsStore.models.length;
  }
  @computed get table(): ITable<IBalanceRow> | null {
    const models: IFinanceActionModel[] = this.financeActionsStore.models;
    if (!models.length) {
      return null;
    }

    const rows: IBalanceRow[] = models.map<IBalanceRow>(model => new BalanceRowDecorator(model));

    const balanceTableBuilder = new TableBuilder<IBalanceRow>();
    const totalRow: IBalanceTotalRow | null = this.getFinanceActionsService.total
      ? new BalanceTotalRowDecorator(this.getFinanceActionsService.total, this.currency)
      : null;

    balanceTableBuilder.setColumnKeys([
      'source',
      'created_at',
      'updated_at',
      'description',
      'action_state',
      'price',
      'profit_diff',
      'total_profit'
    ]);
    balanceTableBuilder.setModels(rows);
    balanceTableBuilder.setFooter(totalRow);

    return balanceTableBuilder.getTable();
  }

  protected financeActionsStore: IStoreWithCurrentModel<IFinanceActionModel>;
  protected getFinanceActionsService: IFinanceActionsDataService;
  protected financeActionsDataHandler: IReactionDisposer;

  protected get currency(): string {
    return 'rub';
  }

  constructor(
    financeActionsStore: IStoreWithCurrentModel<IFinanceActionModel>,
    getFinanceActionsService: IFinanceActionsDataService
  ) {
    this.financeActionsStore = financeActionsStore;
    this.getFinanceActionsService = getFinanceActionsService;

    this.financeActionsDataHandler = reaction<IFinanceActionData[]>(
      () => this.getFinanceActionsService.actions,
      (actions): void => {
        new FinanceActionsStoreInitDecorator(this.financeActionsStore).init(actions);
      }
    );
  }

  @action.bound async init(): Promise<void> {
    const params: IBalanceParams = {};

    await flow(this.getFinanceActionsService.init)(params);
  }

  /**
   * Метод получения списка букингов по передаваемым параметрам
   * @param params Объект с данными для запроса букингов
   */
  @action.bound async getFinanceActions(params: IBalanceParams): Promise<void> {
    await flow(this.getFinanceActionsService.init)(params);
  }

  /**
   * Сбрасывает хранилище
   */
  @action.bound dispose(): void {
    this.financeActionsStore.resetStore();
    this.financeActionsDataHandler();
  }
}
