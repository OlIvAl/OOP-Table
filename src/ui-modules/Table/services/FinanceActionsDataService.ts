import { action, observable, runInAction } from 'mobx';
import { IBalanceParams, IFinanceActionsDataService } from '../state/interfaces';
import {
  IBalanceStateData,
  IFinanceActionData,
  IFinanceActionTotal
} from '../../../DataLayer/modules/actions/interfaces';
import DataLayer from '../../../DataLayer';

export class FinanceActionsDataService implements IFinanceActionsDataService {
  @observable loading: boolean = true;
  @observable actions: IFinanceActionData[] = [];
  @observable total: IFinanceActionTotal | null = null;
  @observable totalAmount: number = 0;

  @action.bound async *init(params: IBalanceParams): AsyncGenerator<void> {
    yield this.setLoading();

    yield this.getData(params);

    yield this.unsetLoading();
  }

  /**
   * Делает запрос за данными списка букингов
   * Инициализирует хранилище списка букингов
   * Инициализирует хранилище програм для фильтрации
   * @param params Объект с данными для запроса букингов
   */
  @action.bound protected async getData(params: IBalanceParams): Promise<void> {
    try {
      const data: IBalanceStateData = await DataLayer.getBalanceStateData(params);

      runInAction(() => {
        this.actions = data.actions;
        this.total = data.total_row;
        this.totalAmount = data.total;
        this.actions = data.actions;
        this.total = data.total_row;
        this.totalAmount = data.total;
      });
    } catch (e) {
      throw e;
    }
  }

  /**
   * Ставит флаг загрузки в true
   */
  @action.bound protected setLoading(): void {
    this.loading = true;
  }
  /**
   * Ставит флаг загрузки в false
   */
  @action.bound protected unsetLoading(): void {
    this.loading = false;
  }
}
