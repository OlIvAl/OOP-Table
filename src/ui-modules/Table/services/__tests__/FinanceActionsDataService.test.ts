import { IFinanceActionsDataService } from '../../state/interfaces';
import { FinanceActionsDataService } from '../FinanceActionsDataService';
import DataLayer from '../../../../DataLayer';


let getFinanceActionsService: IFinanceActionsDataService;
let initGenerator: AsyncGenerator<void>;

beforeEach(
  async (): Promise<void> => {
    getFinanceActionsService = new FinanceActionsDataService();
    initGenerator = getFinanceActionsService.init({});
  }
);

describe('FinanceActionsDataService', () => {
  describe('constructor', () => {
    it('init valid object', () => {
      expect(getFinanceActionsService).toMatchSnapshot();
    });
  });
  describe('init', () => {
    it('set loading', async () => {
      await initGenerator.next();

      expect(getFinanceActionsService.loading).toEqual(true);
    });
    describe('getData', () => {
      it('call getBalanceStateData once', async () => {
        const getBalanceStateDataSpy = jest.spyOn(DataLayer, 'getBalanceStateData');

        await initGenerator.next();
        await initGenerator.next();

        expect(getBalanceStateDataSpy).toHaveBeenCalled();
        expect(getBalanceStateDataSpy).toHaveBeenCalledTimes(1);
      });
      it('throw error, if server send error', async () => {
        const error = Error('');
        jest.spyOn(DataLayer, 'getBalanceStateData').mockImplementation(async () => {
          throw error;
        });

        await initGenerator.next();

        await expect(initGenerator.next()).rejects.toThrow(error);
      });
    });
    it('set total', async () => {
      await initGenerator.next();
      await initGenerator.next();

      expect(getFinanceActionsService.total).toEqual({
        price: 100,
        profit_diff: 1000,
        total_profit: 0
      });
    });
    it('set totalAmount to store', async () => {
      await initGenerator.next();
      await initGenerator.next();

      expect(getFinanceActionsService.totalAmount).toEqual(69);
    });
    it('unset loading', async () => {
      await initGenerator.next();
      await initGenerator.next();
      await initGenerator.next();

      expect(getFinanceActionsService.loading).toEqual(false);
    });
  });
});
