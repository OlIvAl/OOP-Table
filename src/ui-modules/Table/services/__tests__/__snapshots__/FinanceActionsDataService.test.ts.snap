// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinanceActionsDataService constructor init valid object 1`] = `FinanceActionsDataService {}`;
